presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java

```diff
@@ -243,33 +217,117 @@ public MethodsGroupBuilder withExtraParameters(Function<SpecializeContext, List<
             return this;
         }
 
+        public MethodsGroupBuilder blockPositionTypes(String methodName, List<Optional<Class<?>>> types)
+        {
+            Optional<Method> methodOptional = this.methods.stream().filter(method -> method.getName().equals(methodName)).findFirst();
+            checkState(methodOptional.isPresent(), "method needs to be defined first");
+            this.nullConventionAwareMethods.add(new NullConventionAwareMethod(methodOptional.get(), types));
+            return this;
+        }
+
         public MethodsGroup build()
         {
-            return new MethodsGroup(methods, extraParametersFunction);
+            return new MethodsGroup(methods, extraParametersFunction, nullConventionAwareMethods.build());
         }
     }
 
-    static class MethodsGroup
+    public static class ChoiceBuilder
+    {
+        private final Class<?> clazz;
+        private final Signature signature;
+        private boolean nullableResult;
+        private List<ArgumentProperty> argumentProperties;
+        private final ImmutableList.Builder<MethodsGroup> methodsGroups = ImmutableList.builder();
+
+        private ChoiceBuilder(Class<?> clazz, Signature signature)
+        {
+            this.clazz = requireNonNull(clazz, "clazz is null");
+            this.signature = requireNonNull(signature, "signature is null");
+        }
+
+        public ChoiceBuilder implementation(Function<MethodsGroupBuilder, MethodsGroupBuilder> methodsGroupSpecification)
+        {
+            MethodsGroupBuilder methodsGroupBuilder = new MethodsGroupBuilder(clazz);
+            methodsGroupSpecification.apply(methodsGroupBuilder);
+            methodsGroups.add(methodsGroupBuilder.build());
+            return this;
+        }
+
+        public ChoiceBuilder nullableResult(boolean nullableResult)
+        {
+            this.nullableResult = nullableResult;
+            return this;
+        }
+
+        public ChoiceBuilder argumentProperties(ArgumentProperty... argumentProperties)
+        {
+            requireNonNull(argumentProperties, "argumentProperties is null");
+            this.argumentProperties = ImmutableList.copyOf(argumentProperties);
+            return this;
+        }
+
+        public PolymorphicScalarFunctionChoice build()
+        {
+            if (argumentProperties == null) {
+                argumentProperties = nCopies(signature.getArgumentTypes().size(), valueTypeArgumentProperty(RETURN_NULL_ON_NULL));
+            }
+            return new PolymorphicScalarFunctionChoice(nullableResult, argumentProperties, methodsGroups.build());
+        }
+    }
+
+    static final class MethodsGroup
     {
         private final List<Method> methods;
         private final Optional<Function<SpecializeContext, List<Object>>> extraParametersFunction;
+        private final List<NullConventionAwareMethod> nullConventionAwareMethods;
```

Comment:
`methods` and `nullConventionAwareMethods` can probably be merged into one.

```
private final List<MethodAndNativeContainerTypes> methods;
```

`MethodsGroupBuilder.methods` will take care of synthesizing `explicitNativeContainerTypes` as a list of `Optional.empty()`. To do this, you'll need to pass `Signature` to the constructor of `MethodsGroupBuilder` to help you figure out how many elements need to be in the list. Having the `Signature` will also allow you to assert that `methodWithExplicitJavaTypes` is invoked with the correct number of elements in the list.

I'm not sure, but I think this can work.

Reply:
This is a good idea. I merged methods and nullConventionAwareMethods in the new version. Now they are under class MethodAndNativeContainerTypes.